core/src/test/scala/unit/kafka/server/FetchSessionTest.scala

```diff
@@ -877,34 +935,341 @@ class FetchSessionTest {
+    assertPartitionsOrder(context1, partitionsInSession1)
+    val respData1 = new util.LinkedHashMap[TopicIdPartition, FetchResponseData.PartitionData]
+    val errorCode1 = if (startsWithTopicIdsInMetadataCache) Errors.NONE.code else Errors.UNKNOWN_TOPIC_ID.code
+    val fooResponseTp1 = if (startsWithTopicIdsInMetadataCache) tidp0 else nullTidp0
+    val fooResponse1 = if (startsWithTopicIdsInMetadataCache) noErrorResponse else errorResponse(errorCode1)
+    respData1.put(fooResponseTp1, fooResponse1)
+    val resp1 = context1.updateAndGenerateResponseData(respData1)
+    assertEquals(Errors.NONE, resp1.error())
+    assertTrue(resp1.sessionId() != INVALID_SESSION_ID)
+    val responseData1 = resp1.responseData(topicNames, request1.version)
+    assertEquals(errorCode1, responseData1.get(tp0).errorCode)
+
+    // Create an incremental fetch request with an update to the partition.
+    val reqData2 = new util.LinkedHashMap[TopicPartition, FetchRequest.PartitionData]
+    reqData2.put(tp0, new FetchRequest.PartitionData(fooId, 0, 0, 100,
+      Optional.empty()))
+    val request2 = createRequest(new JFetchMetadata(resp1.sessionId(), 1), reqData2, EMPTY_PART_LIST, false)
+    val topicNamesForRequest2 = if (endsWithTopicIdsInMetadataCache) Map(fooId -> "foo").asJava else Map[Uuid, String]().asJava
+    val context2 = fetchManager.newContext(
+      request2.version,
+      request2.metadata,
+      request2.isFromFollower,
+      request2.fetchData(topicNamesForRequest2),
+      request2.forgottenTopics(topicNamesForRequest2),
+      topicNamesForRequest2
+    )
+    assertEquals(classOf[IncrementalFetchContext], context2.getClass)
+    // We will still have the topic ID in the session if we started with topic IDs and currently do not have the ID anymore.
+    val partitionsInSession2 = if (startsWithTopicIdsInMetadataCache || endsWithTopicIdsInMetadataCache) Seq(tidp0) else Seq(nullTidp0)
+    assertPartitionsOrder(context2, partitionsInSession2)
+    val respData2 = new util.LinkedHashMap[TopicIdPartition, FetchResponseData.PartitionData]
+    // If we always had topic IDs in the metadata cache or newly resolved the ID we won't have an error.
+    // If the topic ID was once in the metadata cache and in the session, the partition has likely been deleted and would have an INCONSISTENT_TOPIC_ID error.
+    // Likely if the topic ID was never in the broker, return UNKNOWN_TOPIC_OR_PARTITION
+    val errorCode2 = if (endsWithTopicIdsInMetadataCache) Errors.NONE.code
+    else if (startsWithTopicIdsInMetadataCache) Errors.INCONSISTENT_TOPIC_ID.code else Errors.UNKNOWN_TOPIC_OR_PARTITION.code
+    val fooResponseTp2 = partitionsInSession2(0)
+    val fooResponse2 = if (endsWithTopicIdsInMetadataCache) noErrorResponse else errorResponse(errorCode2)
+    respData2.put(fooResponseTp2, fooResponse2)
+    val resp2 = context2.updateAndGenerateResponseData(respData2)
```

Comment:
I am not sure that I understand the value that we get out of this logic. `updateAndGenerateResponseData` creates a FetchResponse based on its input. Therefore, the response that we assert is not so surprising in the end, right? It will contain `INCONSISTENT_TOPIC_ID` if when the method gets it as an input. This logic would make sense for a test which verifies `updateAndGenerateResponseData` but looks like a distraction in a test which verify the name resolution logic. Am I missing something here?

Reply:
Sure. We can remove it. I think I was concerned about the correct handling of the resolved partitions (ie, we get a response back that we can actually parse), but maybe that's not really necessary.

Reply:
Do you mean the correct handling of the resolved partitions by `updateAndGenerateResponseData`? I think testing `updateAndGenerateResponseData` is a good thing in general. Perhaps, we should just put this into a separate test specific to that method? We should also test it for all context types with and without topic id, I guess.

Reply:
Is there a way to make such a test without duplicating the newContext portions?